pkg/skaffold/build/ko/build.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ko
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/ko/pkg/build"
+	"github.com/google/ko/pkg/publish"
+
+	// latestV1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest/v1"
+	latestV1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/ko/schema"
+)
+
+// Build an artifact using ko
+func (b *Builder) Build(ctx context.Context, out io.Writer, a *latestV1.Artifact, ref string) (string, error) {
+	koBuilder, err := b.newKoBuilder(ctx, a)
+	if err != nil {
+		return "", fmt.Errorf("error creating ko builder: %w", err)
+	}
+
+	koPublisher, err := b.newKoPublisher(ref)
+	if err != nil {
+		return "", fmt.Errorf("error creating ko publisher: %w", err)
+	}
+	defer koPublisher.Close()
+
+	imageRef, err := b.buildAndPublish(ctx, a.ImageName, koBuilder, koPublisher)
+	if err != nil {
+		return "", fmt.Errorf("could not build and publish ko image %q: %w", a.ImageName, err)
+	}
+	fmt.Fprintln(out, imageRef.Name())
+
+	return b.getImageIdentifier(ctx, imageRef, ref)
+}
+
+// buildAndPublish the image using the ko builder and publisher.
+func (b *Builder) buildAndPublish(ctx context.Context, imageName string, koBuilder build.Interface, koPublisher publish.Interface) (name.Reference, error) {
+	importpath, err := getImportPath(imageName, koBuilder)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine Go import path for ko image %q: %w", imageName, err)
+	}
+	imageMap, err := b.publishImages(ctx, []string{importpath}, koPublisher, koBuilder)
```

Comment:
So somewhere you need to pass in the location of the docker daemon.  We normally do this by setting the `DOCKER_HOST` via `b.localDaemon.ExtraEnv()` (e.g., see [jib](https://github.com/GoogleContainerTools/skaffold/blob/5926b5a1d035c601c8bda70998cacbbe4677c690/pkg/skaffold/build/jib/maven.go#L66)).

Reply:
Good catch! Ko creates its own Docker client, which reads `DOCKER_HOST` and other envvars (see https://github.com/google/ko/blob/ee23538378722e060a2f7c7800f226e0b82e09e7/pkg/publish/daemon.go and https://github.com/google/go-containerregistry/blob/ce35c99b3fafc7b6ea9eb192933ae2730fe360d0/pkg/v1/daemon/options.go#L39), but that misses out on Skaffold's parsing of envvars from `minikube docker-env`.

Ko doesn't shell out to the docker client, but rather uses embeds the client lib. We can use `RawClient()` ((https://github.com/GoogleContainerTools/skaffold/blob/586578c7c2abfcee36e2b9a61d031227bb196c2d/pkg/skaffold/docker/image.go#L80)), just like the Buildpacks builder does (https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/buildpacks/lifecycle.go#L123) and pass the client to (https://github.com/google/go-containerregistry/blob/ce35c99b3fafc7b6ea9eb192933ae2730fe360d0/pkg/v1/daemon/options.go#L77).

This will require a bit of plumbing in ko, I'll do that. That will also ensure that we can pass a fake docker client for unit tests.

Reply:
Ko PR to inject docker client: https://github.com/google/ko/pull/378

Reply:
We can remove the blocker on this PR now I think. The ko PR has been merged, and I've updated this PR to pass the Skaffold docker client to the ko builder.